### ISSUE TITLE
Outgoing gRPC/Http2 connection from cdc connector inside docker container hanging

### ISSUE THREAD

**Issue by JanRudert** [GoogleCloudPlatform/DataflowTemplates#331]
This is mainly a copy of this [question I asked on Stackoverflow](https://stackoverflow.com/questions/70663769/outgoing-grpc-http2-from-cdc-connector-inside-docker-container-hanging). Actually it might be an issue with the connector code itself. This is why I create this issue. 

I am running my fork of [Google's cdc connector][1] inside a docker container. 

The connector succesfully connects to postgres and receives the change events. But when trying to lookup the Datacalog entry before sending the first item to Google PubSub it is hanging on the level of Http. The last logs of the embedded netty client are those:

```
[grpc-default-worker-ELG-1-9] DEBUG io.grpc.netty.shaded.io.grpc.netty.NettyClientHandler - [id: 0xd85c91ca, L:/192.168.178.32:54074 - R:datacatalog.googleapis.com/216.58.214.10:443] INBOUND SETTINGS: ack=false settings={MAX_CONCURRENT_STREAMS=100, INITIAL_WINDOW_SIZE=1048576, MAX_HEADER_LIST_SIZE=65536}
[grpc-default-worker-ELG-1-9] DEBUG io.grpc.netty.shaded.io.grpc.netty.NettyClientHandler - [id: 0xd85c91ca, L:/192.168.178.32:54074 - R:datacatalog.googleapis.com/216.58.214.10:443] OUTBOUND SETTINGS: ack=true
[grpc-default-worker-ELG-1-9] DEBUG io.grpc.netty.shaded.io.grpc.netty.NettyClientHandler - [id: 0xd85c91ca, L:/192.168.178.32:54074 - R:datacatalog.googleapis.com/216.58.214.10:443] INBOUND WINDOW_UPDATE: streamId=0 windowSizeIncrement=983041
[grpc-default-worker-ELG-1-9] DEBUG io.grpc.netty.shaded.io.grpc.netty.NettyClientHandler - [id: 0xd85c91ca, L:/192.168.178.32:54074 - R:datacatalog.googleapis.com/216.58.214.10:443] INBOUND SETTINGS: ack=true
```

The docker command I am running is

```
docker run -e "GOOGLE_APPLICATION_CREDENTIALS=/conf/credentials.json" -it --network=host eu.gcr.io/halfpipe-io/nemo-cdc-embedded-connector
```

When I run the application directly and not inside a docker container things work fine. The next Http2 frame logged (following the ones from above) in that case is:

```
[grpc-default-worker-ELG-1-5] DEBUG io.grpc.netty.shaded.io.grpc.netty.NettyClientHandler - [id: 0xbf927477, L:/10.14.128.10:46308 - R:pubsub.googleapis.com/142.250.179.170:443] OUTBOUND HEADERS: streamId=3 headers=GrpcHttp2OutboundHeaders[:authority: pubsub.googleapis.com:443, :path: /google.pubsub.v1.Publisher/Publish, :method: POST, :scheme: https, content-type: application/grpc, te: trailers, user-agent: grpc-java-netty/1.35.0, x-goog-api-client: gl-java/1.8.0_312 gapic/ gax/1.63.0 grpc/, x-goog-request-params: topic=projects/snd-professional/topics/cdc-nemo_angler_staging.public.USERS, grpc-accept-encoding: gzip, authorization: Bearer ya29.c.b0AXv0zTPY71MG6GqeBrQYDBfo92wqXCVg2ullVE0dKsNR-zoOFkabMW9HlT8RmoJE_EVqSltAC0iPosjNdX2Dw0FpyrDyiMVdBQDZPyC2GoU8NsgGlzJeZ9_Sxdv4hHkw3zC5I8y1kOvFB7y1_f44uV1_IcR5slLgy7ySBeTSJuuvjzkmzIEM-7h5cxdL3XQpgBWjYeJvSI0WYzroSZgHTDy-Ede7PfI3sw, grpc-timeout: 4840626u] streamDependency=0 weight=16 exclusive=false padding=0 endStream=false
```

But for some reason this step is never reached inside the container.

I tried to do the same Http call with curl inside the container. This works as well.

```
curl -v --http2 --location "https://datacatalog.googleapis.com/v1beta1/entries:lookup?linkedResource=//pubsub.googleapis.com/projects/<the-project>/topics/<the-topic>" -H "Authorization: Bearer <the-token>"
```

Thanks for any ideas to solve the issue!

  [1]: https://github.com/GoogleCloudPlatform/DataflowTemplates/tree/main/v2/cdc-parent/cdc-embedded-connector

**Comment by JanRudert** [comment]
I have managed to make things working again.

TL;DR the reason was a mixture of outdated libraries and same libraries in different versions being present in the classpath. This revealed only in docker where I suspect the loading order of libs was different.

The details:

I have upgraded the grpc-* libraries version from 1.35.0 to 1.41.0. This was partly achieved by increasing the version of org.apache.beam:beam-sdks-java-bom which has the newer transitive dependencies.This seem to have fixed an issue where the below mentioned IllegalAccessError was stopping the grpc call to work without any logging.

After this upgrade an IllegalAccessError was showing up. It was caused by a conflicting version of com.google.protobuf:protobuf-java. I fixed that by excluding the bad version from io.debezium:debezium-connector-postgres maven dependency.

I suspect there is more potential to clean up dependencies especially looking at the shaded jar from cdc-commons.
